arch/risc-v/src/esp32c3/esp32c3_libc_stubs.c

```diff
@@ -59,12 +60,12 @@ struct _reent;
 
 int _close_r(struct _reent *r, int fd)
 {
-  return close(fd);
+  return nx_close(fd);
```

Comment:
the major difference between close and nx_close is how to handle the error code:

1. nx_close return error code to caller
2. close set error code to errno

Reply:
The main difference is setting `errno` that should not be done when newlib is called from kernel.
@gustavonihei do we expect that newlib stubs are called from the application layer?
For the return codes we can go with
```
int _close_r(struct _reent *r, int fd)
{
  if (nx_close(fd) < 0)
    {
      return ERROR;
    }

  return OK;
}
```

Reply:
> the major difference between close and nx_close is how to handle the error code:
> 
> 1. nx_close return error code to caller
> 2. close set error code to errno
> 
> so, it is no real difference in this case.

> For the return codes we can go with
> 
> ```
> int _close_r(struct _reent *r, int fd)
> {
>   if (nx_close(fd) < 0)
>     {
>       return ERROR;
>     }
> 
>   return OK;
> }
> ```

Seems like a good suggestion.
One thing I just noticed is that `struct reent` provides an `_errno` field, so maybe it could save `nx_close` return value there, similar to how it is being done here: https://chromium.googlesource.com/native_client/nacl-newlib/+/refs/heads/main/newlib/libc/reent/closer.c

Reply:
> of newlib functions which depend on the OS support. One example I can think of is the HMAC support, which motivated this PR some time ago: #7661

Ok. Just let me understand the use case. Till now I thought that libc stubs are called from ESP-IDF layer that runs in kernel, but if the stubs are needed for application usage only, then I will revert changes for libc stubs so they will be mapped to application layer APIs that will lead to a system calls in case of non-flat build mode

Reply:
Regarding `struct reent` and `_errno`, I think it will not help much here as `__getreent` always return `NULL`, so there is no storage to reentrant data.

Reply:
So the main questions are:
1. Is kernel linked with Internal ROM APIs that newlib functions?
2. Are apps linked with Internal ROM APIs that newlib functions?

Those two answers will give us a direction to move forward. If both are "yes", then we will need to use `_NX_CLOSE`, `lib_malloc`, `lib_free`, etc. basically following up how the rest of the things are done in NuttX `libc` implementation.

Reply:
As much as I would like to, unfortunately I cannot dedicate much time to investigate and give you the appropriate answer.
@tmedicci and @lucasssvaz, could you please help on this?

Reply:
Hi! 

> Is kernel linked with Internal ROM APIs that newlib functions?

Yes, there are internal ROM functions that require newlib functions to be defined, including crypto functions. Most of them run on the kernel side (e.g. Wireless), so theoretically, it should have no problem using kernel functions on the defining the stubs.
 
> Are apps linked with Internal ROM APIs that newlib functions?
I can't recall an app that uses them directly, but it'd be safer if we consider that. `hmac_md5` is an example: no app uses it directly, but we can't say for sure that clients don't.

I'd say that is safer to use `_NX_CLOSE`, `lib_malloc`, `lib_free` etc for now.

Reply:
If we are having separate kernel and apps then we need to have two variants of stubs as we are not able to determine on runtime. To use `_NX_CLOSE`, `lib_malloc`, `lib_free` etc. we will need to have 2 stubs structures and app vs kernel to be liked with different variants (like we are having two variants of libc: one for kernel and one for apps). For now let me revert the changes related to stubs until we have a better solution as this seems to be a grey zone for me right now.

Reply:
> To use _NX_CLOSE, lib_malloc, lib_free etc. we will need to have 2 stubs structures and app vs kernel to be liked with different variants (like we are having two variants of libc: one for kernel and one for apps) and this will require the stubs code to be compiled two times with different options

I thought the same, but after a quick review, I think we can continue on this approach: please note that 1) on flat build, the preprocessor will select the `close` and `malloc`-like functions and they can be used for applications without problems (as it was prior this patch). Considering the protected mode, it will use the kernel functions (as expected). Finally, we won't be able to use the functions from the userspace when protected mode is enabled, but this is expected (we should integrate them into the associated interfaces).

Reply:
> Only one thing that I would like to highlight that for example `esp32_rom.ld` provides symbols `PROVIDE ( asprintf = 0x40056d9c );` and `PROVIDE ( strdup = 0x4000143c );` so if kernel code will link with it the memory allocation will most probably be done by newlib via `_malloc_r` implemented in this file and will lead to kernel memory allocation.

The symbols included via `PROVIDE`  keyword behave as weak references, so `asprintf` and `strdup` from NuttX `libkc.a` will prevail in the Kernel image.

Reply:
Done

Reply:
Thanks @pkarashchenko !

(thanks @gustavonihei for the observations ;) )

I'll re-run our internal CI.